File: gcc/scil/tests/RCX/timer.cs
using Hardware;
using Renesas;

namespace Lego
{ 
    public class Timer
    {
        public static Cpu.SchedulerDelegate Scheduler;
    
        public static void Init ()
        {
            Stop ();
            
            Cpu.Scheduler = HandlerA;
            Rcx.VectorTable.Timer16Bit.OutputCompareA = Cpu.InvokeScheduler;
            Rcx.VectorTable.Timer16Bit.OutputCompareB = HandlerB;
          
            H8_3297.Board.Timer16Bit.ControlStatusRegister = 0x01;
            H8_3297.Board.Timer16Bit.ControlRegister = 0x02;

            H8_3297.Board.Timer16Bit.CompareMatchA = 500;
            H8_3297.Board.Timer16Bit.CompareMatchB = 100;
        }
      
        public static void Start ()
        {
            /* Enable interrupt when timer equals comparator A or B. */
            H8_3297.Board.Timer16Bit.InterruptEnableRegister |= 0x08 /* 0x0C */;
        }
      
        public static void Stop ()
        {
            /* Disable interrupt when timer equals comparator A or B. */
            H8_3297.Board.Timer16Bit.InterruptEnableRegister &= ~0x0C;
        }
                  
        public static Cpu.Stack HandlerA (Cpu.Stack stack)
        {
            H8_3297.Board.Timer16Bit.ControlStatusRegister &= ~0x08;
            return Scheduler (stack);
        }

        private static void HandlerB ()
        {
            H8_3297.Board.Timer16Bit.ControlStatusRegister &= ~0x04;
        }
    }
}

File: gcc/scil/tests/c/timer.c
#include "Rcx.h"
#include "Timer.h"

static void HandlerA (void);
static void HandlerB (void);

void 
Lego_Timer_Init ()
{
    Lego_Timer_Stop ();
    
    Lego_Rcx_VectorTable_Timer16Bit_OutputCompareA = HandlerA;
    Lego_Rcx_VectorTable_Timer16Bit_OutputCompareB = HandlerB;
            
    Renesas_H8_3297_Board_Timer16Bit_ControlStatusRegister = 0x01;
    Renesas_H8_3297_Board_Timer16Bit_ControlRegister = 0x02;

    Renesas_H8_3297_Board_Timer16Bit_set_CompareMatchA (500);
    Renesas_H8_3297_Board_Timer16Bit_set_CompareMatchB (100);
}
      
void 
Lego_Timer_Start ()
{
    /* Enable interrupt when timer equals comparator A or B_ */
    Renesas_H8_3297_Board_Timer16Bit_InterruptEnableRegister |= 0x08 /* 0x0C */;
}

void 
Lego_Timer_Stop ()
{
    /* Disable interrupt when timer equals comparator A or B_ */
    Renesas_H8_3297_Board_Timer16Bit_InterruptEnableRegister &= 0xF3;
}

void  __attribute__ ((saveall))
HandlerA ()
{
    Renesas_H8_3297_Board_Timer16Bit_ControlStatusRegister &= 0xF7;
}

void  __attribute__ ((saveall))
HandlerB ()
{
    Renesas_H8_3297_Board_Timer16Bit_ControlStatusRegister &= 0xFB;
}
